src/mlpack/core/tree/cellbound_impl.hpp

```diff
@@ -0,0 +1,970 @@
+}
+
+/**
+ * Same as the copy constructor.
+ */
+template<typename MetricType, typename ElemType>
+inline CellBound<MetricType, ElemType>& CellBound<MetricType, ElemType>::operator=(
+    const CellBound<MetricType, ElemType>& other)
+{
+  if (dim != other.Dim())
+  {
+    // Reallocation is necessary.
+
+    dim = other.Dim();
+    bounds = new math::RangeType<ElemType>[dim];
+  }
+
+  loBound = other.loBound;
+  hiBound = other.hiBound;
+  numBounds = other.numBounds;
+  loAddress = other.loAddress;
+  hiAddress = other.hiAddress;
+
+  // Now copy each of the bound values.
+  for (size_t i = 0; i < dim; i++)
+    bounds[i] = other.bounds[i];
+
+  minWidth = other.MinWidth();
+
+  return *this;
+}
+
+/**
+ * Move constructor: take possession of another bound's information.
+ */
+template<typename MetricType, typename ElemType>
+inline CellBound<MetricType, ElemType>::CellBound(
+    CellBound<MetricType, ElemType>&& other) :
+    dim(other.dim),
+    bounds(other.bounds),
+    loBound(std::move(other.loBound)),
+    hiBound(std::move(other.hiBound)),
+    numBounds(std::move(other.numBounds)),
+    loAddress(std::move(other.loAddress)),
+    hiAddress(std::move(other.hiAddress)),
+    minWidth(other.minWidth)
+{
+  // Fix the other bound.
+  other.dim = 0;
+  other.bounds = NULL;
+  other.minWidth = 0.0;
+}
+
+/**
+ * Destructor: clean up memory.
+ */
+template<typename MetricType, typename ElemType>
+inline CellBound<MetricType, ElemType>::~CellBound()
+{
+  if (bounds)
+    delete[] bounds;
+}
+
+/**
+ * Resets all dimensions to the empty set.
+ */
+template<typename MetricType, typename ElemType>
+inline void CellBound<MetricType, ElemType>::Clear()
+{
+  for (size_t k = 0; k < dim; k++)
+  {
+    bounds[k] = math::RangeType<ElemType>();
+
+    loAddress[k] = std::numeric_limits<AddressElemType>::max();
+    hiAddress[k] = 0;
+  }
+
+  minWidth = 0;
+}
+
+/***
+ * Calculates the centroid of the range, placing it into the given vector.
+ *
+ * @param centroid Vector which the centroid will be written to.
+ */
+template<typename MetricType, typename ElemType>
+inline void CellBound<MetricType, ElemType>::Center(
+    arma::Col<ElemType>& center) const
+{
+  // Set size correctly if necessary.
+  if (!(center.n_elem == dim))
+    center.set_size(dim);
+
+  for (size_t i = 0; i < dim; i++)
+    center(i) = bounds[i].Mid();
+}
+
+template<typename MetricType, typename ElemType>
+void CellBound<MetricType, ElemType>::AddBound(
+    const arma::Col<ElemType>& loCorner,
+    const arma::Col<ElemType>& hiCorner)
+{
+  assert(numBounds < loBound.n_cols);
+  assert(loBound.n_rows == dim);
+  assert(loCorner.n_elem == dim);
+  assert(hiCorner.n_elem == dim);
+
+  // If the subrectangle is not contained entirely in the outer rectangle,
+  // we shrink it.
+  for (size_t k = 0; k < dim; k++)
+  {
+    loBound(k, numBounds) = std::max(loCorner[k], bounds[k].Lo());
+
+    hiBound(k, numBounds) = std::min(bounds[k].Hi(), hiCorner[k]);
```

Comment:
That's true, and I guess in those cases you can remove those hyperrectangles.  I thought about it more... if all of the hyperrectangles are tight in this manner, then is there any reason to store the minimum bounding hyperrectangle?  I don't think it would be useful anymore.  But maybe I have overlooked something.